ngmix/prepsfmom.py

```diff
@@ -0,0 +1,639 @@
+        else:
+            raise ValueError(
+                "The kernel '%s' _PrePSFMom is not recognized!" % self.kernel
+            )
+
+        # compute the total variance from weight map
+        msk = obs.weight > 0
+        tot_var = np.sum(1.0 / obs.weight[msk])
+
+        # run the actual measurements and return
+        mom, mom_cov = _measure_moments_fft(
+            kim, kpsf_im, tot_var, eff_pad_factor, kernels,
+            im_row - psf_im_row, im_col - psf_im_col,
+        )
+        res = _make_mom_res(mom, mom_cov)
+        if res['flags'] != 0:
+            logger.debug("pre-psf moments failed: %s" % res['flagstr'])
+
+        if return_kernels:
+            # put the kernels back into their unpacked state
+            full_kernels = {}
+            for k in kernels:
+                if k == "msk":
+                    continue
+                if k == "nrm":
+                    full_kernels[k] = kernels[k]
+                else:
+                    full_kernels[k] = np.zeros((fft_dim, fft_dim), dtype=np.complex128)
+                    full_kernels[k][kernels["msk"]] = kernels[k]
+            res["kernels"] = full_kernels
+
+        return res
+
+
+class KSigmaMom(_PrePSFMom):
+    """Measure pre-PSF weighted real-space moments w/ the 'ksigma'
+    Fourier-space kernels from Bernstein et al., arXiv:1508.05655.
+
+    If the fwhm of the weight/kernel function is of similar size to the PSF or
+    smaller, then the object properties returned by this fitter will be very noisy.
+
+    Parameters
+    ----------
+    fwhm : float
+        This parameter is the approximate real-space FWHM of the kernel. The units are
+        whatever the Jacobian on the obs converts pixels units to. This is typically
+        arcseconds.
+    pad_factor : int, optional
+        The factor by which to pad the FFTs used for the image. Default is 4.
+    """
+    def __init__(self, fwhm, pad_factor=4):
+        super().__init__(fwhm, kernel='ksigma', pad_factor=pad_factor)
+
+
+class PrePSFGaussMom(_PrePSFMom):
+    """Measure pre-PSF weighted real-space moments w/ a Gaussian kernel.
+
+    This fitter differs from `GaussMom` in that it deconvolves the PSF first.
+
+    If the fwhm of the weight/kernel function is of similar size to the PSF or
+    smaller, then the object properties returned by this fitter will be very noisy.
+
+    Parameters
+    ----------
+    fwhm : float
+        This parameter is the real-space FWHM of the kernel. The units are
+        whatever the Jacobian on the obs converts pixels units to. This is typically
+        arcseconds.
+    pad_factor : int, optional
+        The factor by which to pad the FFTs used for the image. Default is 4.
+    """
+    def __init__(self, fwhm, pad_factor=4):
+        super().__init__(fwhm, kernel='gauss', pad_factor=pad_factor)
+
+
+def _measure_moments_fft(kim, kpsf_im, tot_var, eff_pad_factor, kernels, drow, dcol):
+    # we only need to do things where the ksigma kernel is non-zero
```

Comment:
This comment specifically mentions ksigma

Reply:
fixed